crates/dekaf/src/session.rs

```diff
@@ -300,8 +324,70 @@ impl Session {
         &mut self,
         requests: Vec<messages::metadata_request::MetadataRequestTopic>,
     ) -> anyhow::Result<Vec<MetadataResponseTopic>> {
-        match self.auth.as_ref().unwrap() {
-            SessionAuthentication::Redirect { .. } => {
+        let topics: anyhow::Result<_> = async {
+            let auth = self
+                .auth
+                .as_mut()
+                .ok_or(anyhow::anyhow!("Session not authenticated"))?;
+            let pg_client = &auth.flow_client().await?.pg_client();
+
+            // Re-declare here to drop mutable reference
+            let auth = self.auth.as_ref().unwrap();
+
+            // Concurrently fetch Collection instances for all requested topics.
+            let collections: Vec<(TopicName, Option<Collection>)> = futures::future::try_join_all(
+                requests.clone().into_iter().map(|topic| async move {
+                    Collection::new(
+                        auth,
+                        pg_client,
+                        from_downstream_topic_name(topic.name.to_owned().unwrap_or_default())
+                            .as_str(),
+                    )
+                    .await
+                    .map(|coll| (topic.name.unwrap_or_default(), coll))
+                }),
+            )
+            .await?;
+
+            let mut topics = vec![];
+
+            for (name, maybe_collection) in collections {
+                let Some(collection) = maybe_collection else {
+                    topics.push(
+                        MetadataResponseTopic::default()
+                            .with_name(Some(self.encode_topic_name(name.to_string())?))
+                            .with_error_code(ResponseError::UnknownTopicOrPartition.code()),
+                    );
+                    continue;
+                };
+
+                let partitions = collection
+                    .partitions
+                    .iter()
+                    .enumerate()
+                    .map(|(index, _)| {
+                        messages::metadata_response::MetadataResponsePartition::default()
+                            .with_partition_index(index as i32)
+                            .with_leader_id(messages::BrokerId(1))
```

Comment:
are broker IDs something the client could potentially be caching? Do we need to use distinct values for distinct data-plane FQDN's ?